lua/ide/components/bufferlist/component.lua

```diff
@@ -0,0 +1,247 @@
+					self.open_buf({ fargs = {} })
+				end,
+			})
+			vim.api.nvim_buf_set_keymap(buf, "n", self.config.keymaps.edit_split, "", {
+				silent = true,
+				callback = function()
+					self.open_buf({ fargs = { "split" } })
+				end,
+			})
+			vim.api.nvim_buf_set_keymap(buf, "n", self.config.keymaps.edit_vsplit, "", {
+				silent = true,
+				callback = function()
+					self.open_buf({ fargs = { "vsplit" } })
+				end,
+			})
+			vim.api.nvim_buf_set_keymap(buf, "n", self.config.keymaps.close, "", {
+				silent = true,
+				callback = function()
+					self.close_buf()
+				end,
+			})
+		end
+
+		-- setup autocmds to refresh
+		vim.api.nvim_create_autocmd({ "BufAdd", "BufDelete", "BufWipeout" }, {
+			callback = self.refresh,
+			group = vim.api.nvim_create_augroup("NvimIdeBufferlist", { clear = true }),
+		})
+
+		return buf
+	end
+
+	function self.open()
+		local log = self.logger.logger_from(nil, "Component.open")
+		log.debug("BufferList component opening, workspace %s", vim.api.nvim_get_current_tabpage())
+
+		-- init buf if not already
+		if self.buf == nil then
+			log.debug("buffer does not exist, creating.", vim.api.nvim_get_current_tabpage())
+			self.buf = setup_buffer()
+		end
+
+		log.debug("using buffer %d", self.buf)
+
+		-- initial render
+		self.refresh()
+		return self.buf
+	end
+
+	function self.post_win_create()
+		local log = self.logger.logger_from(nil, "Component.post_win_create")
+		-- setup web-dev-icons highlights if available
+		if pcall(require, "nvim-web-devicons") then
+			for _, icon_data in pairs(require("nvim-web-devicons").get_icons()) do
+				local hl = "DevIcon" .. icon_data.name
+				vim.cmd(string.format("syn match %s /%s/", hl, icon_data.icon))
+			end
+		end
+		-- set highlights for global icon theme
+		icon_set.set_win_highlights()
+		libwin.set_winbar_title(0, "BUFFERS")
+	end
+
+	function self.get_commands()
+		return commands.new(self).get()
+	end
+
+	function self.refresh()
+		local bufs = vim.tbl_map(function(buf)
+			local icon
+			-- use webdev icons if possible
+			if pcall(require, "nvim-web-devicons") then
+				local name = vim.api.nvim_buf_get_name(buf)
+				local ext = vim.fn.fnamemodify(name, ":e:e")
+				icon = require("nvim-web-devicons").get_icon(name, ext, { default = true })
+				if self.kind == "dir" then
+					icon = require("nvim-web-devicons").get_icon("dir")
+				end
+			else
+				if self.kind ~= "dir" then
+					icon = require("ide.icons").global_icon_set.get_icon("File")
+				else
+					icon = require("ide.icons").global_icon_set.get_icon("Folder")
+				end
+			end
+			return {
+				name = get_unique_filename(vim.api.nvim_buf_get_name(buf)),
+				icon = icon,
+				id = buf,
+			}
+		end, get_listed_bufs())
+		self.bufs = bufs
+		self.render()
+	end
+
+	function self.render()
+		local lines = vim.tbl_map(function(buf)
+			return string.format(" %s %s ", buf.icon, buf.name)
+		end, self.bufs or {})
+
+		vim.api.nvim_buf_set_option(self.buf, "modifiable", true)
```

Comment:
We do have a "Buffer" class which makes some of this boiler plate simpler with "write" method. https://github.com/ldelossa/nvim-ide/blob/main/lua/ide/buffers/buffer.lua

You discretion if you want to directly use it or not, i dont feel to strongly.